File: database/database-annotation-processor/src/test/java/ru/tinkoff/kora/database/common/annotation/processor/vertx/VertxParametersTest.java
package ru.tinkoff.kora.database.common.annotation.processor.vertx;

import io.vertx.sqlclient.Tuple;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentMatchers;
import org.mockito.Mockito;
import ru.tinkoff.kora.annotation.processor.common.TestContext;
import ru.tinkoff.kora.application.graph.TypeRef;
import ru.tinkoff.kora.common.Tag;
import ru.tinkoff.kora.database.common.annotation.processor.DbTestUtils;
import ru.tinkoff.kora.database.common.annotation.processor.entity.TestEntityRecord;
import ru.tinkoff.kora.database.common.annotation.processor.vertx.repository.AllowedParametersRepository;
import ru.tinkoff.kora.database.vertx.VertxConnectionFactory;
import ru.tinkoff.kora.database.vertx.mapper.parameter.VertxParameterColumnMapper;
import ru.tinkoff.kora.database.vertx.mapper.result.VertxRowSetMapper;

import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;

public class VertxParametersTest extends AbstractVertxRepositoryTest {
    @Test
    void oldTest() {
        var ctx = new TestContext();
        ctx.addContextElement(TypeRef.of(VertxConnectionFactory.class), executor);
        ctx.addMock(TypeRef.of(VertxRowSetMapper.class, Void.class));
        ctx.addMock(TypeRef.of(VertxParameterColumnMapper.class, TestEntityRecord.UnknownTypeField.class));
        ctx.addMock(TypeRef.of(VertxParameterColumnMapper.class, byte[].class));
        ctx.addMock(TypeRef.of(VertxEntity.TestEntityFieldVertxParameterColumnMapperNonFinal.class));
        var repository = ctx.newInstance(DbTestUtils.compileClass(AllowedParametersRepository.class));
    }

    @Test
    public void testParametersWithSimilarNames() {
        var repository = compileVertx(List.of(), """
            @Repository
            public interface TestRepository extends VertxRepository {
                @Query("INSERT INTO test(value1, value2) VALUES (:value, :valueTest)")
                void test(String value, int valueTest);

            }
            """);

        repository.invoke("test", "test", 42);

        Mockito.verify(executor.connection).preparedQuery(eq("INSERT INTO test(value1, value2) VALUES ($1, $2)"));
        Mockito.verify(executor.query).execute(ArgumentMatchers.argThat(argument -> Tuple.of("test", 42).deepToString().equals(argument.deepToString())), any());
    }

    @Test
    public void testConnectionParameter() {
        var repository = compileVertx(List.of(), """
            @Repository
            public interface TestRepository extends VertxRepository {
                @Query("INSERT INTO test(test) VALUES ('test')")
                void test(SqlConnection connection);
            }
            """);

        repository.invoke("test", executor.connection);

        verify(executor.query).execute(tupleMatcher(Tuple.tuple()), any());
    }

    @Test
    public void testEntityFieldMapping() {
        var repository = compileVertx(List.of(), """
            public final class StringToJsonbParameterMapper implements VertxParameterColumnMapper<String> {
                
                @Override
                public Object apply(String value) {
                    return java.util.Map.of("test", value);
                }
            }
            """, """
            public record SomeEntity(long id, @Mapping(StringToJsonbParameterMapper.class) String value) {}
                
            """, """
            @Repository
            public interface TestRepository extends VertxRepository {
                @Query("INSERT INTO test(id, value) VALUES (:entity.id, :entity.value)")
                void test(SomeEntity entity);
            }
            """);

        repository.invoke("test", newObject("SomeEntity", 42L, "test-value"));

        verify(executor.query).execute(tupleMatcher(Tuple.of(42L, Map.of("test", "test-value"))), any());
    }

    @Test
    public void testNativeParameterWithMapping() {
        var repository = compileVertx(List.of(), """
            public final class StringToJsonbParameterMapper implements VertxParameterColumnMapper<String> {
                
                @Override
                public Object apply(String value) {
                    return java.util.Map.of("test", value);
                }
            }
            """, """
            @Repository
            public interface TestRepository extends VertxRepository {
                @Query("INSERT INTO test(id, value) VALUES (:id, :value)")
                void test(long id, @Mapping(StringToJsonbParameterMapper.class) String value);
            }
            """);

        repository.invoke("test", 42L, "test-value");

        verify(executor.query).execute(tupleMatcher(Tuple.of(42L, Map.of("test", "test-value"))), any());
    }

    @Test
    public void testUnknownTypeParameter() {
        var mapper = Mockito.mock(VertxParameterColumnMapper.class);
        var repository = compileVertx(List.of(mapper), """
            @Repository
            public interface TestRepository extends VertxRepository {
                @Query("INSERT INTO test(id, value) VALUES (:id, :value)")
                void test(long id, UnknownType value);
            }
            """, """
            public class UnknownType {}
            """);

        repository.invoke("test", 42L, newObject("UnknownType"));
    }

    @Test
    public void testUnknownTypeEntityField() {
        var mapper = Mockito.mock(VertxParameterColumnMapper.class);
        var repository = compileVertx(List.of(mapper), """
            @Repository
            public interface TestRepository extends VertxRepository {
                @Query("INSERT INTO test(id, value) VALUES (:id, :value.f)")
                void test(long id, TestEntity value);
            }
            """, """
            public class UnknownType {}
            """, """
            public record TestEntity(UnknownType f){}
            """);

        repository.invoke("test", 42L, newObject("TestEntity", newObject("UnknownType")));
    }

    @Test
    public void testNativeParameterNonFinalMapper() {
        var repository = compileVertx(List.of(newGeneratedObject("TestMapper")), """
            public class TestMapper implements VertxParameterColumnMapper<String> {
                @Override
                public Object apply(String value) {
                    return java.util.Map.of("test", value);
                }
            }
            """, """
            @Repository
            public interface TestRepository extends VertxRepository {
                @Query("INSERT INTO test(id, value) VALUES (:id, :value)")
                void test(long id, @Mapping(TestMapper.class) String value);
            }
            """);

        repository.invoke("test", 42L, "test-value");
    }

    @Test
    public void testMultipleParametersWithSameMapper() {
        var repository = compileVertx(List.of(newGeneratedObject("TestMapper")), """
            public class TestMapper implements VertxParameterColumnMapper<String> {
                @Override
                public Object apply(String value) {
                    return java.util.Map.of("test", value);
                }
            }
            """, """
            @Repository
            public interface TestRepository extends VertxRepository {
                @Query("INSERT INTO test(id, value) VALUES (:id, :value)")
                void test1(long id, @Mapping(TestMapper.class) String value);
                @Query("INSERT INTO test(id, value) VALUES (:id, :value)")
                void test(long id, @Mapping(TestMapper.class) String value);
            }
            """);
    }

    @Test
    public void testMultipleParameterFieldsWithSameMapper() {
        var repository = compileVertx(List.of(newGeneratedObject("TestMapper")), """
            public class TestMapper implements VertxParameterColumnMapper<TestRecord> {
                @Override
                public Object apply(TestRecord value) {
                    return java.util.Map.of("test", value.toString());
                }
            }
            """, """
            @Repository
            public interface TestRepository extends VertxRepository {
                @Query("INSERT INTO test(id, value) VALUES (:id, :value.f1)")
                void test1(long id, TestRecord value);
                @Query("INSERT INTO test(id, value) VALUES (:id, :value.f1)")
                void test2(long id, TestRecord value);
                @Query("INSERT INTO test(id, value1, value2) VALUES (:id, :value1.f1, :value2.f1)")
                void test2(long id, TestRecord value1, TestRecord value2);
            }
            """, """
            public record TestRecord(@Mapping(TestMapper.class) TestRecord f1, @Mapping(TestMapper.class) TestRecord f2){}
            """);
    }

    @Test
    public void testEntityFieldMappingByTag() throws ClassNotFoundException {
        var mapper = Mockito.mock(VertxParameterColumnMapper.class);
        var repository = compileVertx(List.of(mapper), """
            public record SomeEntity(long id, @Tag(SomeEntity.class) String value) {}
                
            """, """
            @Repository
            public interface TestRepository extends VertxRepository {
                @Query("INSERT INTO test(id, value) VALUES (:entity.id, :entity.value)")
                void test(SomeEntity entity);
            }
            """);

        repository.invoke("test", newObject("SomeEntity", 42L, "test-value"));

        verify(mapper).apply("test-value");

        var mapperConstructorParameter = repository.repositoryClass.getConstructors()[0].getParameters()[1];
        assertThat(mapperConstructorParameter.getType()).isEqualTo(VertxParameterColumnMapper.class);
        var tag = mapperConstructorParameter.getAnnotation(Tag.class);
        assertThat(tag).isNotNull();
        assertThat(tag.value()).isEqualTo(new Class<?>[]{compileResult.loadClass("SomeEntity")});
    }

    @Test
    public void testParameterMappingByTag() throws ClassNotFoundException {
        var mapper = Mockito.mock(VertxParameterColumnMapper.class);
        var repository = compileVertx(List.of(mapper), """
            @Repository
            public interface TestRepository extends VertxRepository {
                @Query("INSERT INTO test(value) VALUES (:value)")
                void test(@Tag(TestRepository.class) String value);
            }
            """);

        repository.invoke("test", "test-value");

        verify(mapper).apply("test-value");

        var mapperConstructorParameter = repository.repositoryClass.getConstructors()[0].getParameters()[1];
        assertThat(mapperConstructorParameter.getType()).isEqualTo(VertxParameterColumnMapper.class);
        var tag = mapperConstructorParameter.getAnnotation(Tag.class);
        assertThat(tag).isNotNull();
        assertThat(tag.value()).isEqualTo(new Class<?>[]{compileResult.loadClass("TestRepository")});
    }

    @Test
    public void testRecordParameterMappingByTag() {
        var mapper = Mockito.mock(VertxParameterColumnMapper.class);
        var repository = compileVertx(List.of(mapper), """
            @Repository
            public interface TestRepository extends VertxRepository {
                @Query("INSERT INTO test(value) VALUES (:value)")
                void test(@Tag(TestRepository.class) TestRecord value);
            }
            """, """
            public record TestRecord(String value){}
            """);

        var mapperConstructorParameter = repository.repositoryClass.getConstructors()[0].getParameters()[1];
        assertThat(mapperConstructorParameter.getType()).isEqualTo(VertxParameterColumnMapper.class);
        var tag = mapperConstructorParameter.getAnnotation(Tag.class);
        assertThat(tag).isNotNull();
        assertThat(tag.value()).isEqualTo(new Class<?>[]{compileResult.loadClass("TestRepository")});
    }

}

File: database/database-symbol-processor/src/test/kotlin/ru/tinkoff/kora/database/symbol/processor/vertx/VertxParametersTest.kt
package ru.tinkoff.kora.database.symbol.processor.vertx

import io.vertx.sqlclient.Tuple
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.mockito.kotlin.any
import org.mockito.kotlin.same
import org.mockito.kotlin.verify
import ru.tinkoff.kora.database.vertx.mapper.parameter.VertxParameterColumnMapper

class VertxParametersTest : AbstractVertxRepositoryTest() {
    @Test
    fun testConnectionParameter() {
        val repository = compile(listOf<Any>(), """
            @Repository
            interface TestRepository : VertxRepository {
                @Query("INSERT INTO test(test) VALUES ('test')")
                fun test(session: io.vertx.sqlclient.SqlClient)
            }
            """.trimIndent())

        repository.invoke<Any>("test", executor.connection)

        verify(executor.query).execute(Tuple.tuple().matcher(), any())
    }

    @Test
    fun testNativeParameter() {
        val repository = compile(listOf<Any>(), """
            @Repository
            interface TestRepository : VertxRepository {
                @Query("INSERT INTO test(value1, value2) VALUES (:value1, :value2)")
                fun test(value1: String?, value2: Int)
            }
            
        """.trimIndent())

        repository.invoke<Any>("test", null, 1)
        verify(executor.connection).preparedQuery("INSERT INTO test(value1, value2) VALUES ($1, $2)")
        verify(executor.query).execute(Tuple.of(null, 1).matcher(), any())
        executor.reset()

        repository.invoke<Any>("test", "test", 1)
        verify(executor.connection).preparedQuery("INSERT INTO test(value1, value2) VALUES ($1, $2)")
        verify(executor.query).execute(Tuple.of("test", 1).matcher(), any())
    }


    @Test
    fun testParametersWithSimilarNames() {
        val repository = compile(listOf<Any>(), """
            @Repository
            interface TestRepository : VertxRepository {
                @Query("INSERT INTO test(value1, value2) VALUES (:value, :valueTest)")
                fun test(value: String?, valueTest: Int)
            }
            """.trimIndent())

        repository.invoke<Any>("test", "test", 42)

        verify(executor.connection).preparedQuery("INSERT INTO test(value1, value2) VALUES ($1, $2)")
        verify(executor.query).execute(Tuple.of("test", 42).matcher(), any())
    }

    @Test
    fun testEntityFieldMapping() {
        val repository = compile(listOf<Any>(), """
            class StringToJsonbParameterMapper: VertxParameterColumnMapper<String?> {
                override fun apply(value: String?): Any? {
                    return mapOf("test" to value)
                }
            }
            """.trimIndent(), """
            public data class SomeEntity(val id: Long, @Mapping(StringToJsonbParameterMapper::class) val value: String)
            """.trimIndent(), """
            @Repository
            interface TestRepository: VertxRepository {
                @Query("INSERT INTO test(id, value) VALUES (:entity.id, :entity.value)")
                fun test(entity: SomeEntity)
            }

            """.trimIndent())

        repository.invoke<Any>("test", new("SomeEntity", 42L, "test-value"))

        verify(executor.connection).preparedQuery("INSERT INTO test(id, value) VALUES ($1, $2)")
        verify(executor.query).execute(Tuple.of(42L, mapOf("test" to "test-value")).matcher(), any())
    }

    @Test
    fun testNativeParameterWithMapping() {
        val repository = compile(listOf<Any>(), """
            class StringToJsonbParameterMapper: VertxParameterColumnMapper<String?> {
                override fun apply(value: String?): Any? {
                    return mapOf("test" to value)
                }
            }
            """.trimIndent(), """
            @Repository
            interface TestRepository: VertxRepository {
                @Query("INSERT INTO test(id, value) VALUES (:id, :value)")
                fun test(id: Long, @Mapping(StringToJsonbParameterMapper::class) value: String)
            }
            """.trimIndent())

        repository.invoke<Any>("test", 42L, "test-value")

        verify(executor.connection).preparedQuery("INSERT INTO test(id, value) VALUES ($1, $2)")
        verify(executor.query).execute(Tuple.of(42L, mapOf("test" to "test-value")).matcher(), any())
    }

    @Test
    fun testDataClassParameter() {
        val repository = compile(listOf<Any>(), """
        @Repository
        interface TestRepository: VertxRepository {
            @Query("INSERT INTO test(id, value) VALUES (:entity.id, :entity.value)")
            fun test(entity: TestEntity)
        }
        """.trimIndent(), """
        data class TestEntity(val id: Long, val value: String?)    
        """.trimIndent())

        repository.invoke<Any>("test", new("TestEntity", 42, null))
        verify(executor.connection).preparedQuery("INSERT INTO test(id, value) VALUES ($1, $2)")
        verify(executor.query).execute(Tuple.of(42L, null).matcher(), any())
        executor.reset()

        repository.invoke<Any>("test", new("TestEntity", 42, "test"))
        verify(executor.connection).preparedQuery("INSERT INTO test(id, value) VALUES ($1, $2)")
        verify(executor.query).execute(Tuple.of(42L, "test").matcher(), any())
    }


    @Test
    fun testUnknownTypeParameter() {
        val mapper = Mockito.mock(VertxParameterColumnMapper::class.java) as VertxParameterColumnMapper<Any>
        val repository = compile(listOf(mapper), """
            @Repository
            interface TestRepository : VertxRepository {
                @Query("INSERT INTO test(id, value) VALUES (:id, :value)")
                fun test(id: Long, value: UnknownType)
            }
            
            """.trimIndent(), """
            public class UnknownType {}
            
            """.trimIndent())
        val o = new("UnknownType")

        repository.invoke<Any>("test", 42L, o)

        verify(mapper).apply(same(o))
    }

    @Test
    fun testUnknownTypeEntityField() {
        val mapper = Mockito.mock(VertxParameterColumnMapper::class.java) as VertxParameterColumnMapper<Any>
        val repository = compile(listOf(mapper), """
            @Repository
            interface TestRepository : VertxRepository {
                @Query("INSERT INTO test(id, value) VALUES (:id, :value0.f)")
                fun test(id: Long, value0: TestEntity)
            }
            
            """.trimIndent(), """
            class UnknownType {}
            """.trimIndent(), """
            data class TestEntity (val f: UnknownType)
            """.trimIndent())
        val o = new("UnknownType")

        repository.invoke<Any>("test", 42L, new("TestEntity", o))

        verify(mapper).apply(same(o))
    }

    @Test
    fun testNativeParameterNonFinalMapper() {
        val repository = compile(listOf(newGenerated("TestMapper")), """
            open class TestMapper : VertxParameterColumnMapper<String> {
                override fun apply(value0: String?) =  mapOf("test" to value0)
            }
            """.trimIndent(), """
            @Repository
            interface TestRepository : VertxRepository {
                @Query("INSERT INTO test(id, value) VALUES (:id, :value0)")
                fun test(id: Long, @Mapping(TestMapper::class) value0: String);
            }
            
            """.trimIndent())

        repository.invoke<Any>("test", 42L, "test-value")

        verify(executor.query).execute(Tuple.of(42L, mapOf("test" to "test-value")).matcher(), any())
    }

    @Test
    fun testMultipleParametersWithSameMapper() {
        val repository = compile(listOf(newGenerated("TestMapper")), """
            open class TestMapper : VertxParameterColumnMapper<String> {
                override fun apply(value0: String?) =  mapOf("test" to value0)
            }
            """.trimIndent(), """
            @Repository
            interface TestRepository : VertxRepository {
                @Query("INSERT INTO test(id, value) VALUES (:id, :value0)")
                fun test1(id: Long, @Mapping(TestMapper::class) value0: String);
                @Query("INSERT INTO test(id, value) VALUES (:id, :value0)")
                fun test(id: Long, @Mapping(TestMapper::class) value0: String);
            }
            
            """.trimIndent())
    }

    @Test
    fun testMultipleParameterFieldsWithSameMapper() {
        val repository = compile(listOf(newGenerated("TestMapper")), """
            open class TestMapper : VertxParameterColumnMapper<TestRecord> {
                override fun apply(value0: TestRecord?) =  mapOf("test" to value0.toString())
            }
            """.trimIndent(), """
            @Repository
            interface TestRepository : VertxRepository {
                @Query("INSERT INTO test(id, value) VALUES (:id, :value0.f1)")
                fun test1(id: Long, value0: TestRecord);
                @Query("INSERT INTO test(id, value) VALUES (:id, :value0.f1)")
                fun test2(id: Long, value0: TestRecord);
                @Query("INSERT INTO test(id, value1, value2) VALUES (:id, :value1.f1, :value2.f1)")
                fun test2(id: Long, value1: TestRecord, value2: TestRecord);
            }
            
            """.trimIndent(), """
            data class TestRecord(@Mapping(TestMapper::class) val f1: TestRecord, @Mapping(TestMapper::class) val f2: TestRecord){}
            
            """.trimIndent())
    }
}
